rust/lance-encoding/src/encodings/logical/primitive.rs

```diff
@@ -704,89 +1156,318 @@ impl MiniBlockScheduler {
         };
 
         Ok(Self {
-            meta_buf_position,
-            meta_buf_size,
-            data_buf_position,
+            buffer_offsets_and_sizes: buffer_offsets_and_sizes.to_vec(),
             rep_decompressor: rep_decompressor.into(),
             def_decompressor: def_decompressor.into(),
             value_decompressor: value_decompressor.into(),
+            repetition_index_depth: layout.repetition_index_depth as u16,
             priority,
-            rows_in_page,
+            items_in_page,
             chunk_meta: Vec::new(),
+            rep_index: Vec::new(),
             dictionary,
             def_meaning: def_meaning.into(),
         })
     }
 
-    /// Calculates the overlap between a user-supplied range and a chunk of mini-block data
-    fn calc_overlap(
-        range: &mut Range<u64>,
-        chunk: &ChunkMeta,
-        rows_offset: u64,
-        dst: &mut ScheduledChunk,
-    ) -> ChunkOverlap {
-        if range.start > chunk.num_values + rows_offset {
-            ChunkOverlap::RangeAfterChunk
-        } else {
-            let start_in_chunk = range.start - rows_offset;
-            let end_in_chunk = (start_in_chunk + range.end - range.start).min(chunk.num_values);
-            let rows_in_chunk = end_in_chunk - start_in_chunk;
-            range.start += rows_in_chunk;
-            dst.ranges.push(start_in_chunk..end_in_chunk);
-            ChunkOverlap::Overlap
-        }
+    fn lookup_chunks(&self, chunk_indices: &[usize]) -> Vec<LoadedChunk> {
+        chunk_indices
+            .iter()
+            .map(|&chunk_idx| {
+                let chunk_meta = &self.chunk_meta[chunk_idx];
+                let bytes_start = chunk_meta.offset_bytes;
+                let bytes_end = bytes_start + chunk_meta.chunk_size_bytes;
+                LoadedChunk {
+                    byte_range: bytes_start..bytes_end,
+                    items_in_chunk: chunk_meta.num_values,
+                    chunk_idx,
+                    data: LanceBuffer::empty(),
+                }
+            })
+            .collect()
     }
 }
 
-#[derive(Debug)]
-struct ScheduledChunk {
-    data: LanceBuffer,
-    // The total number of values in the chunk, not all values may be targeted
-    vals_in_chunk: u64,
-    // The number of values that are targeted by the ranges.  This should be the
-    // same as the sum of `Self::ranges`
-    vals_targeted: u64,
-    ranges: Vec<Range<u64>>,
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum PreambleAction {
+    Take,
+    Skip,
+    Absent,
 }
 
-impl Clone for ScheduledChunk {
-    fn clone(&self) -> Self {
-        Self {
-            data: self.data.try_clone().unwrap(),
-            vals_in_chunk: self.vals_in_chunk,
-            ranges: self.ranges.clone(),
-            vals_targeted: self.vals_targeted,
+// TODO: Add test cases for the all-preamble and all-trailer cases
```

Comment:
Is this TODO done? or a follow up?

Reply:
A follow-up.  I'll go through and make follow-ups before I merge, I think there are 2 or 3 in this PR.